### ISSUE TITLE
Nodetool cleanup command failed when cleanup on the same keyspace was running on other node

### ISSUE THREAD

**Issue by Mark-Gurevich** [scylladb/scylladb#13828]
## Issue description

- [ ] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

The nemesis runs cleanup on all nodes in parallel, without specifying keyspaces, so the cleanup  was launched on the **same keyspace** on **different nodes**.

This nodetool error event on node 3:
```
2023-05-03 15:00:10.557: (NodetoolEvent Severity.ERROR) period_type=end event_id=83c999e4-0a9e-45b8-9a6d-f0e97013cacc duration=3m4s: nodetool_command=cleanup node=longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3 errors=["Encountered a bad command exit code!\n\nCommand: '/usr/bin/nodetool  cleanup keyspace1'\n\nExit code: 2\n\nStdout:\n\n\n\nStderr:\n\n\tat sun.rmi.transport.Transport.serviceCall(Transport.java:196)\n\tat sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)\n\tat java.security.AccessController.doPrivileged(Native Method)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:750)\n\n\n", 'Traceback (most recent call last):\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 2554, in run_nodetool\n    self.remoter.run(cmd, timeout=timeout, ignore_status=ignore_status, verbose=verbose, retry=retry)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 613, in run\n    result = _run()\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/decorators.py", line 67, in inner\n    return func(*args, **kwargs)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 604, in _run\n    return self._run_execute(cmd, timeout, ignore_status, verbose, new_session, watchers)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 537, in _run_execute\n    result = connection.run(**command_kwargs)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 620, in run\n    return self._complete_run(channel, exception, timeout_reached, timeout, result, warn, stdout, stderr)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 654, in _complete_run\n    raise UnexpectedExit(result)\nsdcm.remote.libssh2_client.exceptions.UnexpectedExit: Encountered a bad command exit code!\n\nCommand: \'/usr/bin/nodetool  cleanup keyspace1\'\n\nExit code: 2\n\nStdout:\n\n\n\nStderr:\n\n\tat sun.rmi.transport.Transport.serviceCall(Transport.java:196)\n\tat sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)\n\tat java.security.AccessController.doPrivileged(Native Method)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:750)\n\n\n\n']
Traceback (most recent call last):
File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 2554, in run_nodetool
self.remoter.run(cmd, timeout=timeout, ignore_status=ignore_status, verbose=verbose, retry=retry)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 613, in run
result = _run()
File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/decorators.py", line 67, in inner
return func(*args, **kwargs)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 604, in _run
return self._run_execute(cmd, timeout, ignore_status, verbose, new_session, watchers)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 537, in _run_execute
result = connection.run(**command_kwargs)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 620, in run
return self._complete_run(channel, exception, timeout_reached, timeout, result, warn, stdout, stderr)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 654, in _complete_run
raise UnexpectedExit(result)
sdcm.remote.libssh2_client.exceptions.UnexpectedExit: Encountered a bad command exit code!
Command: '/usr/bin/nodetool  cleanup keyspace1'
Exit code: 2
Stdout:
Stderr:
at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)
at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)
at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)
at java.security.AccessController.doPrivileged(Native Method)
at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:750)
```

The error on node 3:
```
2023-05-03 14:57:33.811 <2023-05-03 14:57:33.000>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=54fc672b-8572-464b-b1b3-0ac4b76b30f6: type=RUNTIME_ERROR regex=std::runtime_error line_number=2684 node=longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3
2023-05-03T14:57:33+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3      !ERR | scylla[5513]:  [shard 0] api - Failed force_keyspace_cleanup of keyspace1.table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}: std::runtime_error (cleanup request failed: there is an ongoing cleanup on keyspace1.standard1)
```

Nodetool error event on node 2:
```
2023-05-03 14:58:37.636: (NodetoolEvent Severity.ERROR) period_type=end event_id=3c32ba96-6720-41a5-b391-fbbb482a99f3 duration=1m45s: nodetool_command=cleanup node=longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-2 errors=["Encountered a bad command exit code!\n\nCommand: '/usr/bin/nodetool  cleanup keyspace1'\n\nExit code: 2\n\nStdout:\n\n\n\nStderr:\n\n\tat sun.rmi.transport.Transport.serviceCall(Transport.java:196)\n\tat sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)\n\tat java.security.AccessController.doPrivileged(Native Method)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:750)\n\n\n", 'Traceback (most recent call last):\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 2554, in run_nodetool\n    self.remoter.run(cmd, timeout=timeout, ignore_status=ignore_status, verbose=verbose, retry=retry)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 613, in run\n    result = _run()\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/decorators.py", line 67, in inner\n    return func(*args, **kwargs)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 604, in _run\n    return self._run_execute(cmd, timeout, ignore_status, verbose, new_session, watchers)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 537, in _run_execute\n    result = connection.run(**command_kwargs)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 620, in run\n    return self._complete_run(channel, exception, timeout_reached, timeout, result, warn, stdout, stderr)\n  File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 654, in _complete_run\n    raise UnexpectedExit(result)\nsdcm.remote.libssh2_client.exceptions.UnexpectedExit: Encountered a bad command exit code!\n\nCommand: \'/usr/bin/nodetool  cleanup keyspace1\'\n\nExit code: 2\n\nStdout:\n\n\n\nStderr:\n\n\tat sun.rmi.transport.Transport.serviceCall(Transport.java:196)\n\tat sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)\n\tat java.security.AccessController.doPrivileged(Native Method)\n\tat sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:750)\n\n\n\n']
Traceback (most recent call last):
File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 2554, in run_nodetool
self.remoter.run(cmd, timeout=timeout, ignore_status=ignore_status, verbose=verbose, retry=retry)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 613, in run
result = _run()
File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/decorators.py", line 67, in inner
return func(*args, **kwargs)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 604, in _run
return self._run_execute(cmd, timeout, ignore_status, verbose, new_session, watchers)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/remote_base.py", line 537, in _run_execute
result = connection.run(**command_kwargs)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 620, in run
return self._complete_run(channel, exception, timeout_reached, timeout, result, warn, stdout, stderr)
File "/home/ubuntu/scylla-cluster-tests/sdcm/remote/libssh2_client/__init__.py", line 654, in _complete_run
raise UnexpectedExit(result)
sdcm.remote.libssh2_client.exceptions.UnexpectedExit: Encountered a bad command exit code!
Command: '/usr/bin/nodetool  cleanup keyspace1'
Exit code: 2
Stdout:
Stderr:
at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:573)
at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:834)
at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:688)
at java.security.AccessController.doPrivileged(Native Method)
at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:687)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:750)
```

Error message on node 2:
```
2023-05-03 14:58:05.706 <2023-05-03 14:58:05.000>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=54fc672b-8572-464b-b1b3-0ac4b76b30f6: type=RUNTIME_ERROR regex=std::runtime_error line_number=2856 node=longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-2
2023-05-03T14:58:05+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-2      !ERR | scylla[5376]:  [shard 0] api - Failed force_keyspace_cleanup of keyspace1.table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}: std::runtime_error (cleanup request failed: there is an ongoing cleanup on keyspace1.standard1)
```

*Describe your issue in detail and steps it took to produce it.*

## Impact

*Describe the impact this issue causes to the user.*

## How frequently does it reproduce?

Its new code in SCT to run nodetool cleanup command on all nodes in parallel.

## Installation details

Kernel Version: 5.15.0-1035-aws
Scylla version (or git commit hash): `5.2.0~rc5-20230425.a89867d8c230` with build-id `15657fcbd406469aa63748f0b8aab3330c137b4e`

Cluster size: 3 nodes (i4i.large)

Scylla Nodes used in this run:
- longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3 (34.243.250.166 | 10.4.1.44) (shards: 2)
- longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-2 (54.217.81.169 | 10.4.3.241) (shards: 2)
- longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-1 (34.243.0.143 | 10.4.0.84) (shards: 2)

OS / Image: `ami-0859cdda6c3ecfd84` (aws: eu-west-1)

Test: `longevity-5gb-1h-NodeToolCleanupMonkey-aws-test-parallelObject`
Test id: `b336723d-7076-4600-97f5-8b2d5e992b97`
Test name: `scylla-staging/mark/longevity-5gb-1h-NodeToolCleanupMonkey-aws-test-parallelObject`
Test config file(s):

- [longevity-5gb-1h-NodeToolCleanupMonkey.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/48febfdc4cf473bc3910590307f2e773c5c0bdb7/test-cases/nemesis/longevity-5gb-1h-NodeToolCleanupMonkey.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor b336723d-7076-4600-97f5-8b2d5e992b97`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=b336723d-7076-4600-97f5-8b2d5e992b97)
- Show all stored logs command: `$ hydra investigate show-logs b336723d-7076-4600-97f5-8b2d5e992b97`

## Logs:
- **db-cluster-b336723d.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/db-cluster-b336723d.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/db-cluster-b336723d.tar.gz)
- **sct-runner-events-b336723d.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/sct-runner-events-b336723d.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/sct-runner-events-b336723d.tar.gz)
- **sct-b336723d.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/sct-b336723d.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/sct-b336723d.log.tar.gz)
- **monitor-set-b336723d.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/monitor-set-b336723d.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/monitor-set-b336723d.tar.gz)
- **loader-set-b336723d.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/loader-set-b336723d.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/loader-set-b336723d.tar.gz)
- **parallel-timelines-report-b336723d.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/parallel-timelines-report-b336723d.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/b336723d-7076-4600-97f5-8b2d5e992b97/20230503_150840/parallel-timelines-report-b336723d.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-staging/job/mark/job/longevity-5gb-1h-NodeToolCleanupMonkey-aws-test-parallelObject/2/)
</details>

**Comment by roydahan** [comment]
The issue is mostly to understand if there should be a real limitation for not running a cleanup on two different nodes simultaneously.

**Comment by mykaul** [comment]
> The issue is mostly to understand if there should be a real limitation for not running a cleanup on two different nodes simultaneously.

Not on the same keyspace, that's for sure.

**Comment by Mark-Gurevich** [comment]
> > The issue is mostly to understand if there should be a real limitation for not running a cleanup on two different nodes simultaneously.
> 
> Not on the same keyspace, that's for sure.

If we should not run cleanup on the same keyspace on two different nodes simultaneously, then should I close this issue and re-write the code to run simultaneously per keyspace instead of per node?

**Comment by mykaul** [comment]
> > > The issue is mostly to understand if there should be a real limitation for not running a cleanup on two different nodes simultaneously.
> > 
> > 
> > Not on the same keyspace, that's for sure.
> 
> If we should not run cleanup on the same keyspace on two different nodes simultaneously, then should I close this issue and re-write the code to run simultaneously per keyspace instead of per node?

Perhaps, but I'm not sure I see the use case - we don't ask customers or users to run cleanup in this way - we ask them to run it without specifying the keyspace (by default), so what's the use case to do it this way? An optimization?

**Comment by fgelcer** [comment]
> Perhaps, but I'm not sure I see the use case - we don't ask customers or users to run cleanup in this way - we ask them to run it without specifying the keyspace (by default), so what's the use case to do it this way? An optimization?

that is exactly what we tried to do, but then, we reached the "big" keyspace, where it perhaps takes some time to `cleanup`, and then the message of `cleanup request failed: there is an ongoing cleanup on keyspace1.standard1` happens

**Comment by mykaul** [comment]
What's the next step here? I'm unsure I'm seeing a real issue?

**Comment by fgelcer** [comment]
> What's the next step here? I'm unsure I'm seeing a real issue?

cleanup is supposed to run locally on the node, and drop sstables that are not used by local node anymore (due to cluster resize or reshuffle of tokens), hence running in 2 nodes in parallel, on the same keyspace, should not result in any kind of failure, AFAIU.
the keyspace globally should not be affected by a node running cleanup, hence this is the problem we are having, and `nodetool` is failing in all the nodes simultaneously.

**Comment by fgelcer** [comment]
based on https://docs.scylladb.com/stable/operating-scylla/nodetool-commands/cleanup

**Comment by mykaul** [comment]
I'm reading https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/add-node-to-cluster.html and it says:
Note

To prevent data resurrection, it’s essential to complete cleanup after adding nodes and before any node is decommissioned or removed. However, cleanup may consume significant resources. Use the following guideline to reduce cleanup impact:

Tip 1: When adding multiple nodes, run the cleanup operations after all nodes are added on all nodes but the last one to be added.

Tip 2: Postpone cleanup to low demand hours while ensuring it completes successfully before any node is decommissioned or removed.

Tip 3: Run cleanup one node at a time, reducing overall cluster impact.   <----------- !!!!

Overall, I think the failure you've seen is legit, harmless and 'recoverable' - you'll be able to retry later.

**Comment by fgelcer** [comment]
> > I'm reading https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/add-node-to-cluster.html and it says: Note
> > To prevent data resurrection, it’s essential to complete cleanup after adding nodes and before any node is decommissioned or removed. However, cleanup may consume significant resources. Use the following guideline to reduce cleanup impact:
> > Tip 1: When adding multiple nodes, run the cleanup operations after all nodes are added on all nodes but the last one to be added.
> > Tip 2: Postpone cleanup to low demand hours while ensuring it completes successfully before any node is decommissioned or removed.
> > Tip 3: Run cleanup one node at a time, reducing overall cluster impact. <----------- !!!!
> 
> it is a tip to avoid cluster impact, and not documented as a limitation to the feature.. so if my low demand window is "short" and i can increase the impact on my cluster to run on ALL nodes in parallel, it should be no problem, based on current documentation.. but it doesn't work...

so, should we consider a documentation issue? or is it a real Scylla issue, that is not allowing me to run it in parallel? cause nothing is being said about it being a limitation

> 
> > Overall, I think the failure you've seen is legit, harmless and 'recoverable' - you'll be able to retry later.
> 
> i agree it is supposed to be recoverable...

**Comment by mykaul** [comment]
> > > I'm reading https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/add-node-to-cluster.html and it says: Note
> > > To prevent data resurrection, it’s essential to complete cleanup after adding nodes and before any node is decommissioned or removed. However, cleanup may consume significant resources. Use the following guideline to reduce cleanup impact:
> > > Tip 1: When adding multiple nodes, run the cleanup operations after all nodes are added on all nodes but the last one to be added.
> > > Tip 2: Postpone cleanup to low demand hours while ensuring it completes successfully before any node is decommissioned or removed.
> > > Tip 3: Run cleanup one node at a time, reducing overall cluster impact. <----------- !!!!
> > 
> > 
> > it is a tip to avoid cluster impact, and not documented as a limitation to the feature.. so if my low demand window is "short" and i can increase the impact on my cluster to run on ALL nodes in parallel, it should be no problem, based on current documentation.. but it doesn't work...
> 
> so, should we consider a documentation issue? or is it a real Scylla issue, that is not allowing me to run it in parallel? cause nothing is being said about it being a limitation

It's a real issue in the sense that you can't run in parallel on multiple tables, but that is gracefully handled by Scylla, so I don't see an issue here.
 
> 
> > > Overall, I think the failure you've seen is legit, harmless and 'recoverable' - you'll be able to retry later.
> > 
> > 
> > i agree it is supposed to be recoverable...

Isn't it?

**Comment by fgelcer** [comment]
> > > > I'm reading https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/add-node-to-cluster.html and it says: Note
> > > > To prevent data resurrection, it’s essential to complete cleanup after adding nodes and before any node is decommissioned or removed. However, cleanup may consume significant resources. Use the following guideline to reduce cleanup impact:
> > > > Tip 1: When adding multiple nodes, run the cleanup operations after all nodes are added on all nodes but the last one to be added.
> > > > Tip 2: Postpone cleanup to low demand hours while ensuring it completes successfully before any node is decommissioned or removed.
> > > > Tip 3: Run cleanup one node at a time, reducing overall cluster impact. <----------- !!!!
> > > 
> > > 
> > > it is a tip to avoid cluster impact, and not documented as a limitation to the feature.. so if my low demand window is "short" and i can increase the impact on my cluster to run on ALL nodes in parallel, it should be no problem, based on current documentation.. but it doesn't work...
> > 
> > 
> > so, should we consider a documentation issue? or is it a real Scylla issue, that is not allowing me to run it in parallel? cause nothing is being said about it being a limitation
> 
> It's a real issue in the sense that you can't run in parallel on multiple tables, but that is gracefully handled by Scylla, so I don't see an issue here.
> 
> > > > Overall, I think the failure you've seen is legit, harmless and 'recoverable' - you'll be able to retry later.
> > > 
> > > 
> > > i agree it is supposed to be recoverable...
> 
> Isn't it?

even if it is recoverable... it doesn't change the fact that, either the documentation must be clear about running one node at the time (not because cluster impact, but because Scylla limits running one node at the time), or Scylla should be smart and accept running all nodes at the same time...
i'm not discussing if the failure was fatal, but that we are trying to test different scenarios, that our procedures are not limiting us (it is only a tip after all), but it doesn't work

**Comment by fgelcer** [comment]
@tzach ^^

**Comment by mykaul** [comment]
We can add a comment @ https://docs.scylladb.com/stable/operating-scylla/nodetool-commands/cleanup.html regarding this. I think it's benign. At least I have not seen customers / users complain - the error seems reasonable to me and makes sense.

**Comment by tzach** [comment]
Is the fauilre "just" result of overloaded cluster (as a result fo running more than one cleanup) or more than that (e.g. logic mistake of some kind).

If it the first, we can promote the recomendation to a warning.

**Comment by mykaul** [comment]
> Is the fauilre "just" result of overloaded cluster (as a result fo running more than one cleanup) or more than that (e.g. logic mistake of some kind).

The latter - you should not be running cleanup in parallel on the same keyspace, and this is exactly what the issue is. We are gracefully preventing the user from making this mistake.

**Comment by roydahan** [comment]
Again for the background, in the past we were asked (I think by @avikivity) to run our cleanup in parallel for:
1. reduce the cleanup time in the tests (it can take very long time with big datasets and large clusters).
2. measure the performance of cluster impact.

This task was in our backlog and now when we reached it and tried it, apparently there is a limitation for it.

**Comment by mykaul** [comment]
> Again for the background, in the past we were asked (I think by @avikivity) to run our cleanup in parallel for:
> 
> 1. reduce the cleanup time in the tests (it can take very long time with big datasets and large clusters).
> 2. measure the performance of cluster impact.
> 
> This task was in our backlog and now when we reached it and tried it, apparently there is a limitation for it.

Indeed. The limitation is harmless though.

**Comment by Mark-Gurevich** [comment]
> > Again for the background, in the past we were asked (I think by @avikivity) to run our cleanup in parallel for:
> > 
> > 1. reduce the cleanup time in the tests (it can take very long time with big datasets and large clusters).
> > 2. measure the performance of cluster impact.
> > 
> > This task was in our backlog and now when we reached it and tried it, apparently there is a limitation for it.
> 
> Indeed. The limitation is harmless though.

@fgelcer @roydahan 
So what are the next steps here?
Close the issues as invalid?
Change the note in the docs from tip to warning that there is a limitation to run cleanup on more than one node?
Drop the task for parallel cleanup?

**Comment by mykaul** [comment]
I'd fix the QA code to handle the error, in either:
1. Split the cleanup of tables to the different nodes (they can run in parallel, on different tables)
2. If there's a 'collision', skip that table on that node.

**Comment by roydahan** [comment]
I want to test the cleanup as it should be used in the field / cloud.
Inventing our own procedure of how to run cleanup faster is a risk.

@fgelcer please contact the cloud dev-ops team and try to understand how exactly they ran cleanup (possibly also when just to make sure coverage is aligned on both ends).

**Comment by mykaul** [comment]
@bhalevy - please take a look at this issue:
1. Documentation recommends one node at a time.
2. Automatically (or via automation) running cleanup in parallel means you may 'risk' trying to cleanup the same keyspace on multiple nodes - which is not allowed. How do we expect automation to overcome this?

**Comment by bhalevy** [comment]
> @bhalevy - please take a look at this issue:
> 
> 1. Documentation recommends one node at a time.

We should reconsider that.
As for longer term direction - as we discussed, cleanup should be run automatically by a distributed cluster topology service so they would run optimally without any interactive steps.

> 2. Automatically (or via automation) running cleanup in parallel means you may 'risk' trying to cleanup the same keyspace on multiple nodes - which is not allowed. How do we expect automation to overcome this?

Why isn't it allowed?
This error, `there is an ongoing cleanup on ...` refers to local state in the node, not to cleanup on other nodes.
Please double check that the test doesn't run cleanup concurrently more than once on _the same_ node.

**Comment by mykaul** [comment]
@bhalevy, @raphaelsc - so we need to update our docs, which still recommend 1 node at a time.
@roydahan, @fgelcer - can you ensure the parallel cleanup works as expected? 1 cleanup operation per node, but the newly added node.

**Comment by raphaelsc** [comment]
logs show we're trying cleanup multiple times on the table of node 3, so something is off with the test.

```
$ cat messages.log | grep "force_keyspace_cleanup" 
2023-05-03T13:58:46+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 0] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:03:33+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 0] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:09:07+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 0] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:14:04+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 0] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:20:11+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 1] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:25:33+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 1] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:30:13+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 0] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:36:22+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 1] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:41:53+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 0] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:47:00+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 0] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:52:06+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 0] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:57:09+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3     !INFO | scylla[5513]:  [shard 1] api - force_keyspace_cleanup: keyspace=keyspace1 tables={table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}}
2023-05-03T14:57:33+00:00 longevity-5gb-1h-NodeToolCleanupMon-db-node-b336723d-3      !ERR | scylla[5513]:  [shard 0] api - Failed force_keyspace_cleanup of keyspace1.table{name=standard1, id=eb2a4170-e9b9-11ed-95a4-aabd3d1c2040}: std::runtime_error (cleanup request failed: there is an ongoing cleanup on keyspace1.standard1)
```

**Comment by bhalevy** [comment]
> @bhalevy, @raphaelsc - so we need to update our docs, which still recommend 1 node at a time.
> @roydahan, @fgelcer - can you ensure the parallel cleanup works as expected? 1 cleanup operation per node, but the newly added node.

@mykaul I'm all for updating the docs, but let's commit that only after we test that running cleanups on all nodes, under load, is manageable and doesn't kill performance (as it used to a long time ago).

**Comment by raphaelsc** [comment]
> > @bhalevy, @raphaelsc - so we need to update our docs, which still recommend 1 node at a time.
> > @roydahan, @fgelcer - can you ensure the parallel cleanup works as expected? 1 cleanup operation per node, but the newly added node.
> 
> @mykaul I'm all for updating the docs, but let's commit that only after we test that running cleanups on all nodes, under load, is manageable and doesn't kill performance (as it used to a long time ago).

we need a longevity test for that, as we can only promise it if we're certain that every release can keep it. recently we bumped into a disk space issue during cleanup, and if we had ran cleanup on all nodes simulteanously, the problem would have been much worse. perhaps a more conservatice approach is to parallelize it, but only on a subset carefully chosen (e.g. nodes from different regions).

**Comment by bhalevy** [comment]
Would you think scheduling repair in some complicated way on the cluster is reasonable, or major compaction for that matter?
We should be able to run cleanup in parallel and if we hit issues, we need to fix them and make it work.

**Comment by raphaelsc** [comment]
> Would you think scheduling repair in some complicated way on the cluster is reasonable, or major compaction for that matter? We should be able to run cleanup in parallel and if we hit issues, we need to fix them and make it work.

There's no disagreement here about the goal @bhalevy. My point is that the parallelism doesn't have to be only 1 or all, but also something in between. There's a reason today that we prefer to run major only on a subset of the nodes, which is, to reduce the chances of interfering with user workload. But you got a valid point above, where at the end of the day, cleanup is only another compaction and shouldn't be affecting the user workload any more than regular compaction does. But that's only theory, what we expect. We must verify what happens in practice, and of course, improve things as issues are found. If it causes a regression for some users, then it's a regression. Alternatives must be provided. There's no single recipe for everybody that will satisfy them all, all scenarios.

**Comment by vladzcloudius** [comment]
> It's a real issue in the sense that you can't run in parallel on multiple tables, but that is gracefully handled by Scylla, so I don't see an issue here.

@mykaul and Co.: cleanup is a completely isolated node-local operation. Any 'cleanup' running on a remote node should have zero (!!!) impact on a cleanup on a local node.
AFAIU from the off-line conversation with @raphaelsc QA by mistake ran a 'cleanup' on the same item on the same node - hence the error. And this is indeed expected behavior.

@Mark-Gurevich Please, consider updating the GH issue description since it's very confusing.

Guys, in general, please, try to limit any discussion on the GH issue to scope of that issue.

Don't spam it with discussion of side-topics like new APIs. Open a separate GH issue with an RFC and reference it instead.
What is happening on this GH issue is going to make it totally impossible to read in a year from now.

**Comment by vladzcloudius** [comment]
And BTW, the test that runs cleanups on different nodes and the same/different items in parallel is totally valid, @Mark-Gurevich.

@mykaul the recommendation in our documentation you referenced here (https://github.com/scylladb/scylladb/issues/13828#issuecomment-1547450900) is just a recommendation how reduce the overhead on the cluster. 
It doesn't mean that everybody is going to run it like that hence must be completely ignored by our QA.

**Comment by mykaul** [comment]
@Mark-Gurevich - returning to the original issue - please verify you are not running cleanup more than once on the same node.
